server/sonar-server/src/main/java/org/sonar/server/startup/RegisterMetrics.java

```diff
@@ -20,99 +20,142 @@
    * Used when no plugin is defining Metrics
    */
-  public RegisterMetrics(MeasuresDao measuresDao, QualityGateConditionDao conditionDao) {
-    this(measuresDao, conditionDao, new Metrics[]{});
+  public RegisterMetrics(DbClient dbClient) {
+    this(dbClient, new Metrics[] {});
   }
 
   public void start() {
-    Profiler profiler = Profiler.create(LOG).startInfo("Register metrics");
-    measuresDao.disableAutomaticMetrics();
+    register(concat(CoreMetrics.getMetrics(), getPluginMetrics()));
+  }
 
-    List<Metric> metricsToRegister = newArrayList();
-    metricsToRegister.addAll(CoreMetrics.getMetrics());
-    metricsToRegister.addAll(getMetricsRepositories());
-    register(metricsToRegister);
-    cleanAlerts();
+  void register(Iterable<Metric> metrics) {
+    Profiler profiler = Profiler.create(LOG).startInfo("Register metrics");
+    DbSession session = dbClient.openSession(false);
+    try {
+      save(session, metrics);
+      sanitizeQualityGates(session);
+      session.commit();
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
     profiler.stopDebug();
   }
 
+  private void sanitizeQualityGates(DbSession session) {
+    dbClient.gateConditionDao().deleteConditionsWithInvalidMetrics(session);
+  }
+
+  private void save(DbSession session, Iterable<Metric> metrics) {
+    Map<String, MetricDto> basesByKey = new HashMap<>();
+    for (MetricDto base : from(dbClient.metricDao().selectEnabled(session)).toList()) {
+      basesByKey.put(base.getKey(), base);
+    }
+
+    for (Metric metric : metrics) {
+      MetricDto dto = MetricToDto.INSTANCE.apply(metric);
+      MetricDto base = basesByKey.get(metric.getKey());
+      if (base == null) {
+        // new metric, never installed
+        dbClient.metricDao().insert(session, dto);
+      } else if (!base.isUserManaged()) {
+        // existing metric, update changes. Existing custom metrics are kept without applying changes.
+        dto.setId(base.getId());
+        dbClient.metricDao().update(session, dto);
+      }
+      basesByKey.remove(metric.getKey());
+    }
+
+    for (MetricDto nonUpdatedBase : basesByKey.values()) {
+      if (!nonUpdatedBase.isUserManaged()) {
+        LOG.info("Disable metric {} [{}]", nonUpdatedBase.getShortName(), nonUpdatedBase.getKey());
+        dbClient.metricDao().disableByKey(session, nonUpdatedBase.getKey());
+      }
+    }
+  }
+
   @VisibleForTesting
-  List<Metric> getMetricsRepositories() {
+  List<Metric> getPluginMetrics() {
     List<Metric> metricsToRegister = newArrayList();
     Map<String, Metrics> metricsByRepository = Maps.newHashMap();
-
     for (Metrics metrics : metricsRepositories) {
       checkMetrics(metricsByRepository, metrics);
-      metricsToRegister.addAll(removeExistingUserManagedMetrics(metrics.getMetrics()));
+      metricsToRegister.addAll(metrics.getMetrics());
     }
 
     return metricsToRegister;
   }
 
-  private List<Metric> removeExistingUserManagedMetrics(List<Metric> metrics) {
```

Comment:
It was not removed but refactored. A custom metric is never updated when it already exists. This is implemented in line 96.